stko/optimizers/xtb.py

```diff
@@ -496,3 +514,633 @@ def optimize(self, mol):
+        ----------
+        output_file : :class:`str`
+            Name of xTB output file.
+
+        Returns
+        -------
+        :class:`bool`
+            Returns ``False`` if a negative frequency is present.
+
+        Raises
+        -------
+        :class:`XTBOptimizerError`
+            If the optimization failed.
+
+        :class:`XTBConvergenceError`
+            If the optimization did not converge.
+
+        """
+        if not os.path.exists(output_file):
+            # No simulation has been run.
+            return False
+        # If convergence is achieved, then .xtboptok should exist.
+        if os.path.exists('.xtboptok'):
+            return True
+        elif os.path.exists('NOT_CONVERGED'):
+            raise XTBConvergenceError('Optimization not converged.')
+        else:
+            raise XTBOptimizerError('Optimization failed to complete')
+
+    def _run_xtb(self, xyz, out_file):
+        """
+        Run GFN-xTB.
+
+        Parameters
+        ----------
+        xyz : :class:`str`
+            The name of the input structure ``.xyz`` file.
+
+        out_file : :class:`str`
+            The name of output file with xTB results.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        # Modify the memory limit.
+        if self._unlimited_memory:
+            memory = 'ulimit -s unlimited ;'
+        else:
+            memory = ''
+
+        # Set optimization level and type.
+        optimization = f'--opt {self._opt_level}'
+
+        cmd = (
+            f'{memory} {self._xtb_path} {xyz} '
+            f'--gfnff '
+            f'{optimization} --parallel {self._num_cores} '
+            f'--chrg {self._charge} '
+        )
+
+        with open(out_file, 'w') as f:
+            # Note that sp.call will hold the program until completion
+            # of the calculation.
+            sp.call(
+                cmd,
+                stdin=sp.PIPE,
+                stdout=f,
+                stderr=sp.PIPE,
+                # Shell is required to run complex arguments.
+                shell=True
+            )
+
+    def _run_optimization(self, mol):
+        """
+        Run loop of optimizations on `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        opt_complete : :class:`bool`
+            Returns ``True`` if the calculation is complete and
+            ``False`` if the calculation is incomplete.
+
+        """
+
+        xyz = f'input_structure_ff.xyz'
+        out_file = f'optimization_ff.output'
+        mol.write(xyz)
+        self._run_xtb(xyz=xyz, out_file=out_file)
+
+        # Check if the optimization is complete.
+        output_xyz = 'xtbopt.xyz'
+        opt_complete = self._is_complete(out_file)
+        mol = mol.with_structure_from_file(output_xyz)
```

Comment:
If opt_complete is False, will performing mol.with_structure_from_file() return an error because 'xtbopt.xyz' won't exist?

Reply:
Handled with an exception if not output file is present. I.e. if opt_complete is False, an exception is called.

Reply:
Lovely